components/builder-core/src/metrics.rs

```diff
@@ -24,44 +25,57 @@ pub const APP_NAME: &'static str = "bldr";
 // Statsd Listener Address
 pub const STATS_ENV: &'static str = "HAB_STATS_ADDR";
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Counter {
-    GithubApi,
-    GithubAuthenticate,
-    GithubEvent,
-    GithubInstallationToken,
-    SearchPackages,
+pub type ApiEndpoint = &'static str;
+pub type InstallationId = u32;
+
+// Public Interface
+////////////////////////////////////////////////////////////////////////
+
+/// Metric identifiers will usually be static `str`s, but some may
+/// need to be dynamically-generated `String`s. With a `Cow`, we can
+/// accept either.
+pub type MetricId = Cow<'static, str>;
+
+/// All metrics must implement the Metric trait, as well as one of the
+/// type marker traits (e.g., `CounterMetric`).
+pub trait Metric {
+    /// Generate the metric name to be used
+    fn id(&self) -> MetricId;
 }
 
-// Supported metrics
-#[derive(Debug, Clone)]
-pub enum Gauge {
-    PackageCount,
+/// Marker trait for counter metrics (don't want to accidentally
+/// increment a gauge metric!)
+pub trait CounterMetric {}
+
+/// Increment the given metric by one
+pub fn incr<C>(counter: C)
+    where C: Metric + CounterMetric {
+    match sender().send((MetricType::Counter,
```

Comment:
Where does `sender()` come from here?

Reply:
Little further down... it was one of the original pieces of code in this module I kept.